src/sctools/barcode.py

```diff
@@ -112,3 +113,66 @@ def from_iterable_bytes(cls, iterable, barcode_length):
         """construct an ObservedBarcodeSet from an iterable of bytes barcodes"""
         tbe = TwoBit(barcode_length)
         return cls(Counter(tbe.encode(b) for b in iterable), barcode_length=barcode_length)
+
+
+class ErrorsToCorrectBarcodesMap:
+
+    def __init__(self, errors_to_barcodes):
+        """use a hash map of erroneous_barcode -> true barcode to correct errors in cell barcodes
+
+        :param dict errors_to_barcodes:
+        """
+        if not isinstance(errors_to_barcodes, Mapping):
+            raise TypeError('errors_to_barcodes must be a mapping of erroneous barcodes to correct '
+                            'barcodes, not %a' % type(errors_to_barcodes))
+        self._map = errors_to_barcodes
+
+    def correct_barcode(self, barcode):
+        return self._map[barcode]
+
+    @staticmethod
+    def _prepare_single_base_error_hash_table(iterable):
+        """
+
+        :param Iterable iterable: iterable of string barcodes
+        :return:
+        """
+        error_map = {}
+        for barcode in iterable:
+
+            # include correct barcode
+            error_map[barcode] = barcode
+
+            # include all single-base errors
+            for i, nucleotide in enumerate(barcode):
+                errors = set('ACGTN')
+                errors.discard(nucleotide)
+                for e in errors:
+                    error_map[barcode[:i] + e + barcode[i + 1:]] = barcode
+        return error_map
+
+    @classmethod
+    def single_hamming_errors_from_whitelist(cls, whitelist_file):
+        """factory method to generate instance of class from a file containing "correct" barcodes
+
+        :param str whitelist_file:
+        :return CorrectBarcodeErrorsHash:
+        """
+        with open(whitelist_file, 'r') as f:
+            return cls(cls._prepare_single_base_error_hash_table((line[:-1] for line in f)))
+
+    def correct_bam(self, bam_file, output_bam_file):
+        """correct barcodes in a (potentially unaligned) bamfile, given a whitelist
+
+        :param str bam_file:
+        :param str output_bam_file:
+        """
+        with pysam.AlignmentFile(bam_file, 'rb') as fin, \
+                pysam.AlignmentFile(output_bam_file, 'wb', template=fin) as fout:
+            for alignment in fin:
+                try:
+                    tag = self.correct_barcode(alignment.get_tag('CR'))
+                    alignment.set_tag('CB', tag, 'Z')
```

Comment:
Can you pass these in with their parameter names? I'm not familiar with pysam so I'm not sure what `Z` is

Reply:
Also it's clearer to only include code in the `try` block that you expect to throw the error, which looks like it's `tag = self.correct_barcode(alignment.get_tag('CR'))`

Reply:
the parameters are tag, tag_value, tag_type -- is that helpful? My temptation is to add a comment which might be more informative.

Reply:
Also, I normally agree, but in this case I'm only setting the tag if it is discovered; if I move the set call outside, I need another `if` boolean, and so here I left it inside to improve efficiency.

Reply:
Oh, I see you `pass` if there's a KeyError so if you want to set the tag after it's not necessarily there

Reply:
I think parameter names would make it so that you don't have to add a comment about what each one is, but you could certainly do both